examples/api/lib/material/selection_area/selection_area.1.dart

```diff
@@ -0,0 +1,313 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+/// Flutter code sample for [SelectionArea].
+
+void main() => runApp(const SelectionAreaColorTextRedExampleApp());
+
+class SelectionAreaColorTextRedExampleApp extends StatelessWidget {
+  const SelectionAreaColorTextRedExampleApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
+      ),
+      home: const MyHomePage(title: 'Flutter Demo Home Page'),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key, required this.title});
+
+  final String title;
+
+  @override
+  State<MyHomePage> createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final ContextMenuController _menuController = ContextMenuController();
+  final GlobalKey<SelectionAreaState> selectionAreaKey = GlobalKey<SelectionAreaState>();
+  final Key _text1Id = UniqueKey();
+  final Key _text2Id = UniqueKey();
+  final Key _text3Id = UniqueKey();
+
+  Map<Key, TextSpan> dataSourceMap = <Key, TextSpan>{};
+  Map<Key, TextSpan> bulletSourceMap = <Key, TextSpan>{};
+  late final Map<Key, TextSpan> originSourceData;
+  late final Map<Key, TextSpan> originBulletSourceData;
```

Comment:
Back to my ranting about trying to get rid of these maps... So assuming that we go with the "flattened" approach, where onSelectionChanged is called with 2 indices into a flat string, I think you could take the following approach to this example:

 * The only state you need is a list of ranges representing the parts of the flat string that have been turned red. No need for any of these Maps.
 * Somewhere store your original widget tree under the SelectionListener, with nothing red.
 * Write a method `static Widget _makeRed(Widget child, Iterable<TextRange> ranges);`. It will walk the `child` tree and build a new tree with the given ranges made red. Assert if `ranges` exceeds the number of characters you find in `child`.
 * In your build method: `return _makeRed(selectionListenerChild, redRanges);`
 * onSelectionChanged: `(TextRange selection) { setState(() { redRanges.add(selection); }); }`. Also, you probably would want to merge overlapping ranges.

Would that work?

Reply:
Talked offline, but I think that's reasonable. I worry though that a user would technically have to walk two trees in this scenarios, first the `Widget` tree and then the `TextSpan` tree for the `Text` widgets. Walking one tree seems simpler to me.